phpunit/GLPITestCase.php

```diff
@@ -0,0 +1,337 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+use Glpi\Tests\Log\TestHandler;
+use Monolog\Logger;
+use PHPUnit\Framework\TestCase;
+use Psr\Log\LogLevel;
+
+// Main GLPI test case. All tests should extends this class.
+
+class GLPITestCase extends TestCase
+{
+    private $int;
+    private $str;
+    protected $has_failed = false;
+
+    /**
+     * @var TestHandler
+     */
+    private $php_log_handler;
+
+    /**
+     * @var TestHandler
+     */
+    private $sql_log_handler;
+
+    public function setUp(): void
+    {
+       // By default, no session, not connected
+        $this->resetSession();
+
+       // Ensure cache is clear
+        global $GLPI_CACHE;
+        $GLPI_CACHE->clear();
+
+        // Init log handlers
+        global $PHPLOGGER, $SQLLOGGER;
+        /** @var Monolog\Logger $PHPLOGGER */
+        $this->php_log_handler = new TestHandler(LogLevel::DEBUG);
+        $PHPLOGGER->setHandlers([$this->php_log_handler]);
+        $this->sql_log_handler = new TestHandler(LogLevel::DEBUG);
+        $SQLLOGGER->setHandlers([$this->sql_log_handler]);
+
+        set_error_handler(static function (int $errno, string $errstr) {
+            throw new \Exception($errstr, $errno);
+        }, E_ALL);
+    }
+
+    public function tearDown(): void
+    {
+        restore_error_handler();
+
+        if (isset($_SESSION['MESSAGE_AFTER_REDIRECT']) && !$this->has_failed) {
+            unset($_SESSION['MESSAGE_AFTER_REDIRECT'][INFO]);
+            $this->assertSame(
+                [],
+                $_SESSION['MESSAGE_AFTER_REDIRECT'],
+                sprintf(
+                    "Some messages has not been handled in %s::%s:\n%s",
+                    static::class,
+                    __METHOD__/*$method*/,
+                    print_r($_SESSION['MESSAGE_AFTER_REDIRECT'], true)
+                )
+            );
+        }
+
+        if (!$this->has_failed) {
+            foreach ([$this->php_log_handler, $this->sql_log_handler] as $log_handler) {
+                $this->assertEmpty(
+                    $log_handler->getRecords(),
+                    sprintf(
+                        "Unexpected entries in log in %s::%s:\n%s",
+                        static::class,
+                        __METHOD__/*$method*/,
+                        print_r(array_column($log_handler->getRecords(), 'message'), true)
+                    )
+                );
+            }
+        }
```

Comment:
I was sure I answered this one... But seems like I did dnot hit the send button ^^

I've tested adding a log from `Toolbox::logInfo()` and this results in a tests failure. Sound like this works as expected then, no?

Reply:
`Toolbox::logInfo()` sends log directly in the logger, so it will indeed work. But logs will never be generated by errors as long as the error handler defined by `set_error_handler` convert them to an exception.

Reply:
OK, so I have no solution. Problem is phpunit no longer permit to tests errors,maybe the proper solution would be to use Exceptions GLPI side, and handle properly those that are non fatal - adding an entry in logs anyway that can still be checked from this part.

Reply:
Fixed by last commit.